### PR TITLE
Add migration script

### PR DESCRIPTION
**Pull request by Stell0**
https://github.com/nethesis/dev/issues/5832

### CODE REVIEW

File: root/etc/e-smith/events/actions/nethserver-tancredi-migration [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NethServer.  If not, see COPYING.
+#
+
+if [[ $(/sbin/e-smith/config getprop nethvoice ProvisioningEngine) != 'tancredi' ]]; then
+    echo "Provisioning engine is not tancredi. Don't migrate."
+    exit
+fi
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-tancredi-migration]: This is not an esmith action. As it is intended to run interactively from a shell, it should be moved to /usr/sbin

```suggestion
if [[ $(/sbin/e-smith/config getprop nethvoice ProvisioningEngine) == 'tancredi' ]]; then
    echo "Provisioning engine is already Tancredi. Nothing to do."
    exit 0
fi
/sbin/e-smith/config setprop nethvoice ProvisioningEngine tancredi
```

File: root/etc/e-smith/events/actions/nethserver-tancredi-migration [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+secret=$(perl -mNethServer::Password -e "print NethServer::Password::store('nethvoice')")
+User="admin"
+fpbxPasswordHash=$(fpbx_query "SELECT password_sha1 FROM ampusers WHERE username = '${User}'")
+SecretKey=$(echo -n "${User}${fpbxPasswordHash}${secret}" | sha1sum | awk '{print $1}')
+
+# Get all models available in Tancredi
+TANCREDI_MODELS=($(/usr/bin/curl -s 'http://localhost/tancredi/api/v1/models' -H "User: ${User}" -H "Secretkey: ${SecretKey}" | jq -r '.[].name | @sh'))
+
+# Cicle through all brands known by Tancredi
+for MAC in $(fpbx_query "SELECT mac FROM endpointman_mac_list WHERE mac LIKE '0C383E%' OR mac LIKE '7C2F80%' OR mac LIKE '589EC6%' OR mac LIKE '005058%' OR mac LIKE '000413%' OR mac LIKE '001565%' OR mac LIKE '805E0C%' OR mac LIKE '805EC0%' OR mac LIKE '9C7514%'"); do
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-tancredi-migration]: You could write the WHERE clause similar to

    SUBSTR(mac, 0, 6) IN ('7C2F80', ...)

File: root/etc/e-smith/events/actions/nethserver-tancredi-migration [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+    # Get old provisioning brand
+    BRAND=$(fpbx_query "SELECT endpointman_brand_list.name FROM endpointman_model_list JOIN endpointman_mac_list ON endpointman_model_list.id = endpointman_mac_list.model JOIN endpointman_brand_list ON endpointman_brand_list.id = endpointman_model_list.brand WHERE endpointman_mac_list.mac = '$MAC'")
+
+    # Get old provisioning model
+    MODEL=$(fpbx_query "SELECT endpointman_model_list.model FROM endpointman_model_list JOIN endpointman_mac_list ON endpointman_model_list.id = endpointman_mac_list.model WHERE endpointman_mac_list.mac = '$MAC'")
+
+    # Translate MAC Address from old format to new format
+    TANCREDI_MAC=$(echo $MAC | sed 's/\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)/\1-\2-\3-\4-\5-\6/')
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-tancredi-migration]: A command pipe can be avoided sometimes with redirection

```suggestion
    TANCREDI_MAC=$(sed 's/\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)\([A-F0-9][A-F0-9]\)/\1-\2-\3-\4-\5-\6/' <<<"$MAC")
```

File: migration.php [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (DavidePrincipi) [inline migration.php]: Remove trailing empty lines

```suggestion
```

File: migration.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        JOIN endpointman_brand_list ON endpointman_brand_list.id = endpointman_model_list.brand
+        WHERE SUBSTRING(endpointman_mac_list.mac,1,6) IN ('0C383E','7C2F80','589EC6','005058','000413','001565','805E0C','805EC0','9C7514')";
+$sth = $db->prepare($sql);
+$sth->execute(array());
+$res = $sth->fetchAll(\PDO::FETCH_ASSOC);
+$logger->debug(json_encode($res));
+
+# Cycle through old phones
+foreach ($res as $phone) {
+    $logger->debug("Migrating {$phone['brand']} {$phone['oldmodel']} {$phone['mac']}");
```
Comment (DavidePrincipi) [inline migration.php]: Do not log at this point...

```suggestion
```

File: migration.php [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    # Add phones to tancredi
+    $scope = new \Tancredi\Entity\Scope($mac, $storage, $logger);
+    $scope->metadata['displayName'] = $brand;
+    $scope->metadata['inheritFrom'] = $model;
+    $scope->metadata['scopeType'] = 'phone';
+    $scope->setVariables($variables);
+    \Tancredi\Entity\TokenManager::createToken(uniqid($prefix = rand(), $more_entropy = TRUE), $mac , TRUE); // create first time access token
+    \Tancredi\Entity\TokenManager::createToken(uniqid($prefix = rand(), $more_entropy = TRUE), $mac , FALSE); // create token
+    $phone_scope = \Tancredi\Entity\Scope::getPhoneScope($mac, $storage, $logger);
+
```
Comment (DavidePrincipi) [inline migration.php]:
```suggestion
    $logger->info("Added {$phone['brand']} model {$model} ({$phone['mac']}) from model {$phone['oldmodel']}.");
```

File: migration.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    # Configure RPS with Falconieri
+    setFalconieriRPS($mac, $phone_scope['provisioning_url1']);
```
Comment (DavidePrincipi) [inline migration.php]: I'd like to have the same behavior of UI. If Falconieri fails it shows a warning.

Can we check the Falconieri response status and log a warning if it fails?

**Review by DavidePrincipi — COMMENTED** [review]
I don't get where migration.php is installed...

Do you want to move migration.php to the scripts/ directory?

**Comment by Stell0** [comment]
> I don't get where migration.php is installed...
> 
> Do you want to move migration.php to the scripts/ directory?

yes, done

File: migration.php (hunk 7 above)
Comment (Stell0) [reply migration.php]: Falconieri error_log output is printed to standard output. We could also print it to Tancredi log.

**Review by DavidePrincipi — COMMENTED** [review]
The Bash script is an executable file. I'd avoid the ugly .sh extension for a command. I propose to rename

`tancredi_migration_helper.sh => tancredi-run-migration`

File: migration.php (hunk 7 above)
Comment (DavidePrincipi) [reply migration.php]: I'd expect the log is sent (also) to stdout

File: nethserver-tancredi.spec [hunk 8]
```diff
@@ -48,6 +48,7 @@ perl createlinks
 install NethVoiceAuth.php %{buildroot}/usr/share/tancredi/src/Entity/
 install AsteriskRuntimeFilter.php %{buildroot}/usr/share/tancredi/src/Entity/
 install migration.php  %{buildroot}/usr/share/tancredi/scripts/
+mkdir -p %{buildroot}/usr/sbin
```
Comment (DavidePrincipi) [inline nethserver-tancredi.spec]: If you create it before `%{genfilelist}` the resulting RPM will own `/usr/sbin`: very bad!

```suggestion
```

File: tancredi-migration-helper [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
+        migrate
+        /sbin/e-smith/signal-event nethserver-dnsmasq-update
+    elif [[ $@ =~ -d ]] ; then
+        find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
+        /sbin/e-smith/config setprop nethvoice ProvisioningEngine freepbx
+        /sbin/e-smith/signal-event nethserver-dnsmasq-update
+    else
+        echo "Provisionong engine is already Tancredi"
+        exit 1
```
Comment (DavidePrincipi) [inline tancredi-migration-helper]:
```suggestion
        echo "[NOTICE] Provisionong engine is already Tancredi. Nothing to do."
        exit 0
```

**Review by DavidePrincipi — COMMENTED** [review]
1. It is possible to refactor the IF block to avoid the nested structure
2. I'd just test the first argument, $1 instead of matching among all possible arguments sequences
3. Add `-h` for help?
4. Always write to stdout what has been done - use `[NOTICE]` or `[ERROR]` prefix consistently 

About 1 e 2:

```
IF NOT ENGINE is tancredi

ELIF $1 == -f

ELIF $1 == -d

[ELIF $1 == -h || $1 == --help]

ELSE
    Nothing to do
FI
```

File: tancredi-migration-helper [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    su -s /bin/bash -c "/usr/bin/scl enable rh-php56 -- php /usr/share/tancredi/scripts/migration.php $LK $SECRET" - apache
+}
+
+ENGINE=$(/sbin/e-smith/config getprop nethvoice ProvisioningEngine)
+
+if [[ $ENGINE != 'tancredi' ]] ; then
+    echo "[NOTICE] Migrating provisioning engine to Tancredi"
+    /sbin/e-smith/config setprop nethvoice ProvisioningEngine tancredi
+    migrate
+    /sbin/e-smith/signal-event nethserver-dnsmasq-update
```
Comment (DavidePrincipi) [inline tancredi-migration-helper]:
```suggestion
    /sbin/e-smith/config setprop nethvoice ProvisioningEngine tancredi
    migrate
    /sbin/e-smith/signal-event nethserver-dnsmasq-update
    echo "[NOTICE] Migration completed. The provisioning service is now Tancredi."
```

File: tancredi-migration-helper [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+elif [[ $1 == '-f' ]] ; then
+    echo "[NOTICE] Forcing migration engine to Tancredi"
+    find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
+    migrate
+    /sbin/e-smith/signal-event nethserver-dnsmasq-update
```
Comment (DavidePrincipi) [inline tancredi-migration-helper]:
```suggestion
    echo "[NOTICE] Discarding any existing data in Tancredi database"
    find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
    migrate
    /sbin/e-smith/signal-event nethserver-dnsmasq-update
    echo "[NOTICE] Migration completed in force (-f) mode. The provisioning service is still Tancredi."
```

File: tancredi-migration-helper [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+elif [[ $1 == '-d' ]] ; then
+    echo "[NOTICE] restoriingg freepbx provisioning engine End Point Manager"
+    find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
+    /sbin/e-smith/config setprop nethvoice ProvisioningEngine freepbx
+    /sbin/e-smith/signal-event nethserver-dnsmasq-update
```
Comment (DavidePrincipi) [inline tancredi-migration-helper]:
```suggestion
    find /var/lib/tancredi/data/{scopes,first_access_tokens,tokens} -type f -delete
    /sbin/e-smith/config setprop nethvoice ProvisioningEngine freepbx
    /sbin/e-smith/signal-event nethserver-dnsmasq-update
    echo "[NOTICE] The provisioning engine is now FreePBX End Point Manager."
```

File: tancredi-migration-helper [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+elif [[ $1 == '-h' ]] ; then
+    echo "Usage: $0 [OPTION]"
+    echo -e "Migrate provisioning engine and configured devices to Tancredi\n"
+    echo -e "With no option, migration is launched only once\n"
+    echo -e "-f\t force migration even if tancredi is already the provisioning engine"
+    echo -e "-d\t restore freepbx provisioning engine End Point Manager"
```
Comment (DavidePrincipi) [inline tancredi-migration-helper]:
```suggestion
    echo "Usage: $0 [-h|-f|-d]"
    echo "Import known devices in the Tancredi database, "
    echo "then disable the TFTP service for FreePBX provisoning"
    echo "With no option, the procedure runs once. Accepted options"
    echo -e "  -h\t This help"
    echo -e "  -f\t Drop the Tancredi DB and import known devices from FreePBX again"
    echo -e "  -d\t Restore TFTP configuration for FreePBX provisioning"
```

**Review by DavidePrincipi — APPROVED** [review]

**Pull request merged**